### PR TITLE
Restart unhealthy tasks

### LINKED ISSUE #876: Nomad Restart of Unhealthy Containers
Assuming that `nomad` is configured with a `check` block while registering with `consul` so that Consul does a health check of the containers. 
In this scenario, if consul reports that one of the containers are not healthy, will `nomad` restart/reschedule those containers?

### PR DESCRIPTION
**Pull request by schmichael**
Fixes #876 

Unhealthy checks can now restart tasks. See docs in PR for intended behavior.

- [x] Code
- [x] Docs
- [x] Changelog
- [x] *Tests*

### CODE REVIEW

**Comment by samart** [comment]
I would vote to make grace period just grace period, not including interval.  more straightforward.  How long you want to wait before checking is very different from what interval you want once you are ready.

File: client/task_runner.go [hunk 1]
```diff
@@ -1674,6 +1693,25 @@ func (r *TaskRunner) Restart(source, reason string) {
 	}
 }
 
+// RestartBy deadline. Restarts a task iff the last time it was started was
+// before the deadline.  Returns true if restart occurs; false if skipped.
+func (r *TaskRunner) RestartBy(deadline time.Time, source, reason string) {
```
Comment (jippi) [inline client/task_runner.go]: it doesn't return anything?

**Comment by alxark** [comment]
should we wait for this PR to be merged before next release? it's really very important feature and i'm sure it's blocking some users to start using docker.

**Comment by schmichael** [comment]
> should we wait for this PR to be merged before next release? it's really very important feature and i'm sure it's blocking some users to start using docker.

We decided kind of the opposite. :) *Because* this is such an important feature we didn't want to rush it into the 0.6.1 point release. The delay since was caused by me being on vacation, but rest assured this feature is my highest priority.

File: client/consul_template.go [hunk 2]
```diff
@@ -439,7 +439,8 @@ func (tm *TaskTemplateManager) handleTemplateRerenders(allRenderedTime time.Time
 				}
 
 				if restart {
-					tm.config.Hooks.Restart(consulTemplateSourceName, "template with change_mode restart re-rendered")
+					const failure = false
+					tm.config.Hooks.Restart(consulTemplateSourceName, "template with change_mode restart re-rendered", failure)
```
Comment (schmichael) [inline client/consul_template.go]: My little const pattern is pretty funky, and I'd be happy to remove it.

I do it because I hate seeing method calls with literal booleans in them and having no idea what those booleans do without looking at the method signature and/or docs.

File: client/task_runner_test.go [hunk 3]
```diff
@@ -1251,8 +1269,7 @@ func TestTaskRunner_Template_NewVaultToken(t *testing.T) {
 	})
 
 	// Error the token renewal
-	vc := ctx.tr.vaultClient.(*vaultclient.MockVaultClient)
-	renewalCh, ok := vc.RenewTokens[token]
+	renewalCh, ok := ctx.vault.RenewTokens[token]
```
Comment (schmichael) [inline client/task_runner_test.go]: Sorry for the unrelated test changes. Just made the mocks easier to access.

File: command/agent/consul/catalog_testing.go [hunk 4]
```diff
@@ -25,3 +27,119 @@ func (m *MockCatalog) Service(service, tag string, q *api.QueryOptions) ([]*api.
 	m.logger.Printf("[DEBUG] mock_consul: Service(%q, %q, %#v) -> (nil, nil, nil)", service, tag, q)
 	return nil, nil, nil
 }
+
+// MockAgent is a fake in-memory Consul backend for ServiceClient.
+type MockAgent struct {
```
Comment (schmichael) [inline command/agent/consul/catalog_testing.go]: Not new; moved from `unit_test.go` and exported for use in `client/`

File: command/agent/consul/check_watcher.go [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+		c.unhealthyStart = now
+	}
+
+	// restart timeLimit after start of this check becoming unhealthy
+	restartAt := c.unhealthyStart.Add(c.timeLimit)
+
+	// Must test >= because if limit=1, restartAt == first failure
+	if now.Equal(restartAt) || now.After(restartAt) {
+		// hasn't become healthy by deadline, restart!
+		c.logger.Printf("[DEBUG] consul.health: restarting alloc %q task %q due to unhealthy check %q", c.allocID, c.taskName, c.checkName)
```
Comment (schmichael) [inline command/agent/consul/check_watcher.go]: INFO or WARN level? Left at DEBUG since Task Events are probably a more accessible source for this information.

File: command/agent/consul/unit_test.go [hunk 6]
```diff
@@ -99,126 +148,21 @@ func (t *testFakeCtx) syncOnce() error {
...
 	return &testFakeCtx{
 		ServiceClient: NewServiceClient(fc, true, testLogger()),
 		FakeConsul:    fc,
 		Task:          testTask(),
+		Restarter:     newFakeCheckRestarter(),
 		execs:         make(chan int, 100),
 	}
 }
 
-// fakeConsul is a fake in-memory Consul backend for ServiceClient.
```
Comment (schmichael) [inline command/agent/consul/unit_test.go]: Exported as MockAgent

File: command/agent/consul/client.go [hunk 7]
```diff
@@ -748,15 +778,28 @@ func (c *ServiceClient) UpdateTask(allocID string, existing, newTask *structs.Ta
...
+
 				for _, checkID := range newCheckIDs {
 					sreg.checkIDs[checkID] = struct{}{}
+
 				}
+
+			}
+
+			// Update all watched checks as CheckRestart fields aren't part of ID
+			if check.Watched() {
```
Comment (dadgar) [inline command/agent/consul/client.go]: Get a test ensuring this

File: command/agent/consul/unit_test.go [hunk 8]
```diff
@@ -99,126 +148,21 @@ func (t *testFakeCtx) syncOnce() error {
...
-	c.checkTTLs[id]++
-	return nil
-}
-
 func TestConsul_ChangeTags(t *testing.T) {
 	ctx := setupFake()
 
 	allocID := "allocid"
-	if err := ctx.ServiceClient.RegisterTask(allocID, ctx.Task, nil, nil); err != nil {
+	if err := ctx.ServiceClient.RegisterTask(allocID, ctx.Task, ctx.Restarter, nil, nil); err != nil {
```
Comment (dadgar) [inline command/agent/consul/unit_test.go]: Where are any of the tests ensuring that the watch gets created?

File: command/agent/consul/check_watcher_test.go [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+
+	// Assert Restart was never called
+	if n := len(restarter1.restarts); n > 0 {
+		t.Errorf("expected check 1 to not be restarted but found %d", n)
+	}
+	if n := len(restarter2.restarts); n > 0 {
+		t.Errorf("expected check 2 to not be restarted but found %d", n)
+	}
+}
+
```
Comment (dadgar) [inline command/agent/consul/check_watcher_test.go]: Test checking that Watch() with updated check works

File: command/agent/consul/check_watcher.go [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+// Unwatch a task.
+func (w *checkWatcher) Unwatch(cid string) {
+	c := checkRestart{
+		checkID: cid,
+		remove:  true,
+	}
+	select {
+	case w.watchCh <- &c:
```
Comment (dadgar) [inline command/agent/consul/check_watcher.go]: create an unwatchCh and just pass the cid?

File: command/agent/consul/check_watcher.go [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+
+	// timer for check polling
+	checkTimer := time.NewTimer(0)
+	defer checkTimer.Stop() // ensure timer is never leaked
+
+	// Main watch loop
+	for {
+		// Don't start watching until we actually have checks that
+		// trigger restarts.
+		for len(checks) == 0 {
```
Comment (dadgar) [inline command/agent/consul/check_watcher.go]: Lets just create a single select block. We can start the timer when we get a watch: https://github.com/hashicorp/nomad/blob/master/client/alloc_runner_health_watcher.go#L447

File: command/agent/consul/check_watcher.go [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+// Run the main Consul checks watching loop to restart tasks when their checks
+// fail. Blocks until context is canceled.
+func (w *checkWatcher) Run(ctx context.Context) {
+	defer close(w.done)
+
+	// map of check IDs to their metadata
+	checks := map[string]*checkRestart{}
+
+	// timer for check polling
+	checkTimer := time.NewTimer(0)
```
Comment (dadgar) [inline command/agent/consul/check_watcher.go]: A select on this will fire immediately. You need to stop it first and select from the channel

File: command/agent/consul/check_watcher.go [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+				if c.remove {
+					// should not happen
+					w.logger.Printf("[DEBUG] consul.health: told to stop watching an unwatched check: %q", c.checkID)
+				} else {
+					checks[c.checkID] = c
+					w.logger.Printf("[DEBUG] consul.health: watching alloc %q task %q check %q", c.allocID, c.taskName, c.checkName)
+
+					// Begin polling
+					if !checkTimer.Stop() {
+						<-checkTimer.C
```
Comment (dadgar) [inline command/agent/consul/check_watcher.go]: Do this with a select with a default

File: command/agent/consul/check_watcher.go [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+		return
+	}
+
+	if c.unhealthyStart.IsZero() {
+		// First failure, set restart deadline
+		if c.timeLimit != 0 {
+			c.logger.Printf("[DEBUG] consul.health: alloc %q task %q check %q became unhealthy. Restarting in %s if not healthy",
+				c.allocID, c.taskName, c.checkName, c.timeLimit)
+		}
+		c.unhealthyStart = now
```
Comment (dadgar) [inline command/agent/consul/check_watcher.go]: Can you rename unhealthyStart to unhealthyCheck?

File: command/agent/consul/check_watcher.go [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+	checkName string
+
+	// remove this checkID (if true only checkID will be set)
+	remove bool
+
+	task           TaskRestarter
+	restartDelay   time.Duration
+	grace          time.Duration
+	interval       time.Duration
+	timeLimit      time.Duration
```
Comment (dadgar) [inline command/agent/consul/check_watcher.go]: document non-obvious fields: interval, timeLimit

File: command/agent/consul/check_watcher.go (hunk 13 above)
Comment (schmichael) [reply command/agent/consul/check_watcher.go]: This is straight from the docs: https://golang.org/pkg/time/#Timer.Reset

File: command/agent/consul/check_watcher.go (hunk 14 above)
Comment (schmichael) [reply command/agent/consul/check_watcher.go]: At no other point do I refer to a single run of a check as a check in this file, so I'm not sure I understand the renaming. We don't even have visibility into individual runs of a check since we poll statuses at a fixed interval.

File: client/consul.go [hunk 16]
```diff
@@ -10,8 +10,8 @@ import (
 // ConsulServiceAPI is the interface the Nomad Client uses to register and
 // remove services and checks from Consul.
 type ConsulServiceAPI interface {
-	RegisterTask(allocID string, task *structs.Task, exec driver.ScriptExecutor, net *cstructs.DriverNetwork) error
+	RegisterTask(allocID string, task *structs.Task, restarter consul.TaskRestarter, exec driver.ScriptExecutor, net *cstructs.DriverNetwork) error
 	RemoveTask(allocID string, task *structs.Task)
-	UpdateTask(allocID string, existing, newTask *structs.Task, exec driver.ScriptExecutor, net *cstructs.DriverNetwork) error
+	UpdateTask(allocID string, existing, newTask *structs.Task, restart consul.TaskRestarter, exec driver.ScriptExecutor, net *cstructs.DriverNetwork) error
```
[the remark this answers is not in the document]
Comment (schmichael) [reply client/consul.go]: I was thinking the same thing!

File: command/agent/consul/check_watcher.go (hunk 10 above)
Comment (schmichael) [reply command/agent/consul/check_watcher.go]: That would create a race between adding and removing watches since select cases are randomly selected.

Instead I'll make `watchCh -> watchUpdateCh` and create a small `checkRestartUpdate` struct to handle the add vs remove case.

File: command/agent/consul/check_watcher.go (hunk 13 above)
Comment (schmichael) [reply command/agent/consul/check_watcher.go]: ...but I always forget this caveat on Stop:
> assuming the program has not received from t.C already

Oof this API gets me every time.

File: command/agent/consul/check_watcher.go (hunk 12 above)
Comment (schmichael) [reply command/agent/consul/check_watcher.go]: Fixed below

File: command/agent/consul/unit_test.go (hunk 8 above)
Comment (schmichael) [reply command/agent/consul/unit_test.go]: Added to TestConsul_ChangeChecks and TestConsul_RegServices

File: client/restarts.go [hunk 17]
```diff
@@ -67,15 +69,20 @@ func (r *RestartTracker) SetWaitResult(res *dstructs.WaitResult) *RestartTracker
...
 	defer r.lock.Unlock()
 	r.waitRes = res
+	r.failure = true
 	return r
 }
 
 // SetRestartTriggered is used to mark that the task has been signalled to be
 // restarted
-func (r *RestartTracker) SetRestartTriggered() *RestartTracker {
+func (r *RestartTracker) SetRestartTriggered(failure bool) *RestartTracker {
```
Comment (dadgar) [inline client/restarts.go]: Comment on the param

File: client/restarts_test.go [hunk 18]
```diff
@@ -99,7 +99,7 @@ func TestClient_RestartTracker_RestartTriggered(t *testing.T) {
 	p := testPolicy(true, structs.RestartPolicyModeFail)
 	p.Attempts = 0
 	rt := newRestartTracker(p, structs.JobTypeService)
-	if state, when := rt.SetRestartTriggered().GetState(); state != structs.TaskRestarting && when != 0 {
+	if state, when := rt.SetRestartTriggered(false).GetState(); state != structs.TaskRestarting && when != 0 {
```
Comment (dadgar) [inline client/restarts_test.go]: Unit test the new case

File: client/restarts.go [hunk 19]
```diff
@@ -134,66 +142,40 @@ func (r *RestartTracker) GetState() (string, time.Duration) {
...
-		if r.policy.Mode == structs.RestartPolicyModeFail {
-			r.reason = fmt.Sprintf(
-				`Exceeded allowed attempts %d in interval %v and mode is "fail"`,
-				r.policy.Attempts, r.policy.Interval)
-			return structs.TaskNotRestarting, 0
-		} else {
-			r.reason = ReasonDelay
-			return structs.TaskRestarting, r.getDelay()
+	// Handle restarts due to failures
+	if r.failure {
```
Comment (dadgar) [inline client/restarts.go]: Invert it:

```
if !r.failure {
  return "", 0
}
...
```

File: client/restarts.go [hunk 20]
```diff
@@ -134,66 +142,40 @@ func (r *RestartTracker) GetState() (string, time.Duration) {
...
-	if r.count > r.policy.Attempts {
-		if r.policy.Mode == structs.RestartPolicyModeFail {
-			r.reason = fmt.Sprintf(
-				`Exceeded allowed attempts %d in interval %v and mode is "fail"`,
-				r.policy.Attempts, r.policy.Interval)
-			return structs.TaskNotRestarting, 0
-		} else {
-			r.reason = ReasonDelay
-			return structs.TaskRestarting, r.getDelay()
+		if r.count > r.policy.Attempts {
```
Comment (dadgar) [inline client/restarts.go]: Can you comment how you get to this case.

File: client/task_runner.go [hunk 21]
```diff
@@ -772,7 +789,8 @@ OUTER:
 					return
 				}
 			case structs.VaultChangeModeRestart:
-				r.Restart("vault", "new Vault token acquired")
+				const failure = false
```
Comment (dadgar) [inline client/task_runner.go]: Would it make more sense for the const to be something like `noFailure`

File: command/agent/job_endpoint.go [hunk 22]
```diff
@@ -704,6 +704,13 @@ func ApiTaskToStructsTask(apiTask *api.Task, structsTask *structs.Task) {
 						Header:        check.Header,
 						Method:        check.Method,
 					}
+					if check.CheckRestart != nil {
+						structsTask.Services[i].Checks[j].CheckRestart = &structs.CheckRestart{
```
Comment (dadgar) [inline command/agent/job_endpoint.go]: Capture this in the test in job_endpoint_test.go

File: nomad/structs/structs.go [hunk 23]
```diff
@@ -2757,6 +2757,40 @@ func (tg *TaskGroup) GoString() string {
...
+	return nc
+}
+
+func (c *CheckRestart) Validate() error {
+	if c == nil {
+		return nil
+	}
+
+	if c.Limit < 0 {
+		return fmt.Errorf("limit must be greater than or equal to 0 but found %d", c.Limit)
```
Comment (dadgar) [inline nomad/structs/structs.go]: Use multierror and add a test

File: website/source/docs/job-specification/check_restart.html.md [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+  </tr>
+  <tr>
+    <th width="120">Placement</th>
+    <td>
+      <code>job -> group -> task -> service -> check -> **check_restart**</code>
+    </td>
+  </tr>
+</table>
+
+As of Nomad 0.7 the `check_restart` stanza instructs Nomad when to restart
```
Comment (dadgar) [inline website/source/docs/job-specification/check_restart.html.md]: Mention the merge behavior

File: website/source/docs/job-specification/check_restart.html.md [hunk 25]
```diff
@@ -0,0 +1,151 @@
...
+          name     = "check_table"
+          command  = "/usr/local/bin/check_mysql_table_status"
+          args     = ["--verbose"]
+          interval = "60s"
+          timeout  = "5s"
+
+          check_restart {
+            limit = 3
+            grace = "90s"
+
```
Comment (dadgar) [inline website/source/docs/job-specification/check_restart.html.md]: Remove space

File: website/source/docs/job-specification/check_restart.html.md [hunk 26]
```diff
@@ -0,0 +1,151 @@
...
+        }
+      }
+    }
+  }
+}
+```
+
+- `limit` `(int: 0)` - Restart task when a health check has failed `limit`
+  times.  For example 1 causes a restart on the first failure. The default,
+  `0`, disables healtcheck based restarts. Failures must be consecutive. A
```
Comment (dadgar) [inline website/source/docs/job-specification/check_restart.html.md]: healtcheck. Can you run a spell check on the doc sections

File: website/source/docs/job-specification/check_restart.html.md [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+```hcl
+restart {
+  # ...
+  delay    = "10s"
+  # ...
+}
+```
+
+The [`restart` stanza][restart_stanza] controls the restart behavior of the
+task. In this case it will wait 10 seconds before restarting. Note that even if
```
Comment (dadgar) [inline website/source/docs/job-specification/check_restart.html.md]: `In this case it will stop the task and then wait 10 seconds before starting it again.`.
Delete subsequent sentence

File: command/agent/consul/check_watcher.go [hunk 28]
```diff
@@ -0,0 +1,317 @@
...
+
+	select {
+	case w.checkUpdateCh <- update:
+		// sent watch
+	case <-w.done:
+		// exited; nothing to do
+	}
+}
+
+// Unwatch a task.
```
Comment (dadgar) [inline command/agent/consul/check_watcher.go]: Unwatch a check

File: command/agent/consul/check_watcher.go [hunk 29]
```diff
@@ -0,0 +1,317 @@
...
+					if _, ok := restartedTasks[check.taskKey]; ok {
+						delete(checks, cid)
+					}
+				}
+			}
+		}
+	}
+}
+
+// Watch a task and restart it if unhealthy.
```
Comment (dadgar) [inline command/agent/consul/check_watcher.go]: Watch a check and restart it's task if it becomes unhealthy

File: command/agent/job_endpoint.go (hunk 22 above)
Comment (schmichael) [reply command/agent/job_endpoint.go]: Added https://github.com/hashicorp/nomad/pull/3105/commits/924813d57e06ca5bf6510c998a950f3118900fc7

File: client/restarts_test.go (hunk 18 above)
Comment (schmichael) [reply client/restarts_test.go]: Added https://github.com/hashicorp/nomad/pull/3105/commits/cde908e35d8b3dedc2ed68d2d7b2eb832aaf06ad

File: website/source/docs/job-specification/check_restart.html.md (hunk 24 above)
Comment (schmichael) [reply website/source/docs/job-specification/check_restart.html.md]: Added https://github.com/hashicorp/nomad/pull/3105/commits/10ae18c7eb434af5f57b21a3db3d2acc11328433

File: website/source/docs/job-specification/check_restart.html.md [hunk 30]
```diff
@@ -30,6 +30,8 @@ unhealthy for the `limit` specified in a `check_restart` stanza, it is
 restarted according to the task group's [`restart` policy][restart_stanza]. The
 `check_restart` settings apply to [`check`s][check_stanza], but may also be
 placed on [`service`s][service_stanza] to apply to all checks on a service.
+`check_restart` settings on `service` will only overwrite unset `check_restart`
```
Comment (dadgar) [inline website/source/docs/job-specification/check_restart.html.md]: If `check_restart` is set on both the check and service, the stanza's are merged with the check values taking precedence.

**Comment by jippi** [comment]
Amazing @schmichael !
